File: Corte 1/retoPOO3.py
# Reto 3
# 1) Encapsulamiento en Persona: atributos privados (__) + getters/setters.
# 2) Subclase Empleado que hereda de Persona y agrega atributos y calcularHonorarios().

class Persona:
    def __init__(self):
        self.__tipoDoc = ""
        self.__documento = ""
        self.__nombre = ""
        self.__apellido = ""
        self.__peso = 0.0
        self.__estatura = 0.0
        self.__edad = 0
        self.__sexo = ""

    # ---- Getters y setters (encapsulamiento) ----
    def getTipoDoc(self): return self.__tipoDoc
    def setTipoDoc(self, v): self.__tipoDoc = v
    def getDocumento(self): return self.__documento
    def setDocumento(self, v): self.__documento = v
    def getNombre(self): return self.__nombre
    def setNombre(self, v): self.__nombre = v
    def getApellido(self): return self.__apellido
    def setApellido(self, v): self.__apellido = v
    def getPeso(self): return self.__peso
    def setPeso(self, v): self.__peso = v
    def getEstatura(self): return self.__estatura
    def setEstatura(self, v): self.__estatura = v
    def getEdad(self): return self.__edad
    def setEdad(self, v): self.__edad = v
    def getSexo(self): return self.__sexo
    def setSexo(self, v): self.__sexo = v

    def pedirDatos(self):
        self.__tipoDoc = input("Tipo de documento: ")
        self.__documento = input("Numero de documento: ")
        self.__nombre = input("Nombre: ")
        self.__apellido = input("Apellido: ")
        self.__peso = float(input("Peso (kg): "))
        self.__estatura = float(input("Estatura (m): "))
        self.__edad = int(input("Edad: "))
        self.__sexo = input("Sexo: ")

    def mostrarPersona(self):
        print("---- Datos de la persona ----")
        print(f"Tipo de documento: {self.__tipoDoc}")
        print(f"Documento: {self.__documento}")
        print(f"Nombre: {self.__nombre} {self.__apellido}")
        print(f"Peso: {self.__peso} kg")
        print(f"Estatura: {self.__estatura} m")
        print(f"Edad: {self.__edad}")
        print(f"Sexo: {self.__sexo}")

    def calcularImc(self):
        pesoActual = self.__peso / (self.__estatura ** 2)
        if pesoActual < 20:
            return "PESOBAJO"
        elif pesoActual <= 25:
            return "PESOIDEAL"
        else:
            return "SOBREPESO"

    def mayorEdad(self):
        return self.__edad >= 18


class Empleado(Persona):
    def __init__(self):
        super().__init__()          # inicializa los atributos de Persona
        self.__cargo = ""
        self.__valorHora = 0.0
        self.__horasTrabajadas = 0.0
        self.__departamento = ""

    # ---- Getters y setters propios ----
    def getCargo(self): return self.__cargo
    def setCargo(self, v): self.__cargo = v
    def getValorHora(self): return self.__valorHora
    def setValorHora(self, v): self.__valorHora = v
    def getHorasTrabajadas(self): return self.__horasTrabajadas
    def setHorasTrabajadas(self, v): self.__horasTrabajadas = v
    def getDepartamento(self): return self.__departamento
    def setDepartamento(self, v): self.__departamento = v

    def pedirDatosEmpleado(self):
        self.pedirDatos()                                  # datos heredados de Persona
        self.__cargo = input("Cargo: ")
        self.__valorHora = float(input("Valor por hora: "))
        self.__horasTrabajadas = float(input("Horas trabajadas: "))
        self.__departamento = input("Departamento: ")

    def calcularHonorarios(self):
        total = self.__valorHora * self.__horasTrabajadas
        reteica = total * 0.00966        # 0.966% sobre el valor total
        return total - reteica

    def mostrarEmpleado(self):
        total = self.calcularHonorarios()
        print("---- Datos del empleado ----")
        # usa getters heredados para acceder a los atributos privados de Persona
        print(f"Tipo y numero de documento: {self.getTipoDoc()} {self.getDocumento()}")
        print(f"Nombres y apellidos: {self.getNombre()} {self.getApellido()}")
        print(f"Cargo: {self.__cargo}")
        print(f"Horas trabajadas: {self.__horasTrabajadas}")
        print(f"Valor por hora: {self.__valorHora}")
        print(f"Total a pagar: {total}")


class Inicio:
    @staticmethod
    def main():
        emp = Empleado()
        emp.pedirDatosEmpleado()
        emp.mostrarEmpleado()

        resultado = emp.calcularImc()
        if resultado == "PESOBAJO":
            print("El peso esta por debajo de lo ideal")
        elif resultado == "PESOIDEAL":
            print("El peso es ideal")
        else:
            print("Tiene sobrepeso")

        if emp.mayorEdad():
            print("Es mayor de edad")
        else:
            print("Es menor de edad")


if __name__ == "__main__":
    Inicio.main()

File: Corte 1/Retos/retoPOO3.cpp
#include <iostream>
#include <string>
#include <iomanip>
using namespace std;

// 1) Encapsulamiento: atributos privados + getters/setters publicos.
class Persona {
private:
    string tipoDoc, documento, nombre, apellido, sexo;
    double peso, estatura;
    int edad;

public:
    Persona() : peso(0), estatura(0), edad(0) {}

    // ---- Getters y setters ----
    string getTipoDoc() { return tipoDoc; }
    void setTipoDoc(string v) { tipoDoc = v; }
    string getDocumento() { return documento; }
    void setDocumento(string v) { documento = v; }
    string getNombre() { return nombre; }
    void setNombre(string v) { nombre = v; }
    string getApellido() { return apellido; }
    void setApellido(string v) { apellido = v; }
    double getPeso() { return peso; }
    void setPeso(double v) { peso = v; }
    double getEstatura() { return estatura; }
    void setEstatura(double v) { estatura = v; }
    int getEdad() { return edad; }
    void setEdad(int v) { edad = v; }
    string getSexo() { return sexo; }
    void setSexo(string v) { sexo = v; }

    void pedirDatos() {
        cout << "Tipo de documento: "; getline(cin, tipoDoc);
        cout << "Numero de documento: "; getline(cin, documento);
        cout << "Nombre: "; getline(cin, nombre);
        cout << "Apellido: "; getline(cin, apellido);
        cout << "Peso (kg): "; cin >> peso;
        cout << "Estatura (m): "; cin >> estatura;
        cout << "Edad: "; cin >> edad;
        cin.ignore();
        cout << "Sexo: "; getline(cin, sexo);
    }

    void mostrarPersona() {
        cout << "---- Datos de la persona ----" << endl;
        cout << "Tipo de documento: " << tipoDoc << endl;
        cout << "Documento: " << documento << endl;
        cout << "Nombre: " << nombre << " " << apellido << endl;
        cout << "Peso: " << peso << " kg" << endl;
        cout << "Estatura: " << estatura << " m" << endl;
        cout << "Edad: " << edad << endl;
        cout << "Sexo: " << sexo << endl;
    }

    string calcularImc() {
        double pesoActual = peso / (estatura * estatura);
        if (pesoActual < 20) return "PESOBAJO";
        else if (pesoActual <= 25) return "PESOIDEAL";
        else return "SOBREPESO";
    }

    bool mayorEdad() {
        return edad >= 18;
    }
};

// 2) Empleado hereda de Persona (herencia publica).
class Empleado : public Persona {
private:
    string cargo, departamento;
    double valorHora, horasTrabajadas;

public:
    Empleado() : valorHora(0), horasTrabajadas(0) {}

    string getCargo() { return cargo; }
    void setCargo(string v) { cargo = v; }
    double getValorHora() { return valorHora; }
    void setValorHora(double v) { valorHora = v; }
    double getHorasTrabajadas() { return horasTrabajadas; }
    void setHorasTrabajadas(double v) { horasTrabajadas = v; }
    string getDepartamento() { return departamento; }
    void setDepartamento(string v) { departamento = v; }

    void pedirDatosEmpleado() {
        pedirDatos();                         // datos heredados de Persona
        cout << "Cargo: "; getline(cin, cargo);
        cout << "Valor por hora: "; cin >> valorHora;
        cout << "Horas trabajadas: "; cin >> horasTrabajadas;
        cin.ignore();
        cout << "Departamento: "; getline(cin, departamento);
    }

    double calcularHonorarios() {
        double total = valorHora * horasTrabajadas;
        double reteica = total * 0.00966;     // 0.966% sobre el valor total
        return total - reteica;
    }

    void mostrarEmpleado() {
        double total = calcularHonorarios();
        cout << fixed << setprecision(2);
        cout << "---- Datos del empleado ----" << endl;
        cout << "Tipo y numero de documento: " << getTipoDoc() << " " << getDocumento() << endl;
        cout << "Nombres y apellidos: " << getNombre() << " " << getApellido() << endl;
        cout << "Cargo: " << cargo << endl;
        cout << "Horas trabajadas: " << horasTrabajadas << endl;
        cout << "Valor por hora: " << valorHora << endl;
        cout << "Total a pagar: " << total << endl;
    }
};

int main() {
    Empleado emp;
    emp.pedirDatosEmpleado();
    emp.mostrarEmpleado();

    string resultado = emp.calcularImc();
    if (resultado == "PESOBAJO") cout << "El peso esta por debajo de lo ideal" << endl;
    else if (resultado == "PESOIDEAL") cout << "El peso es ideal" << endl;
    else cout << "Tiene sobrepeso" << endl;

    if (emp.mayorEdad()) cout << "Es mayor de edad" << endl;
    else cout << "Es menor de edad" << endl;

    return 0;
}